### PR TITLE
DSET-4449: Do not count dropped buffer as processed

### PR DESCRIPTION
**Pull request by martin-majlis-s1**
Jira Link: <https://sentinelone.atlassian.net/browse/DSET-4449>

# 🥅 Goal

In the PR #47 I have added counting for dropped and broken buffers. However I haven't adjusted the counting for processed buffers. So when the buffer was dropped we have counted it twice - `enqueued < processed + dropped`.

# 🛠️ Solution

I have refactored the function `DatasetClient.logStatistics()` into two functions - `DatasetClient.Statistics()`, which count the statistics,  and `DatasetClient.logStatistics()`, which logs statistics.

So after fixing that bug it allowed me to write tests to verify that the statistics are computed correctly.

# 🏫 Testing

Add unit tests.

### CODE REVIEW

File: pkg/client/statistics.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package client
+
+import (
+	"time"
+)
+
+type QueueStats struct {
```
Comment (tomaz-s1) [inline pkg/client/statistics.go:7]: A comment on what each of those attributes represents would be nice.

File: pkg/client/client.go [hunk 2]
```diff
@@ -431,54 +432,102 @@ func (client *DataSetClient) logStatistics() {
...
-		zap.Uint64("waiting", eEnqueued-eProcessed-eDropped-eBroken),
-		zap.Float64("processingS", processingInSec),
-	)
+
+	eventsStats := QueueStats{
+		eProcessed,
+		eEnqueued,
+		eDropped,
+		eBroken,
+		eEnqueued - eProcessed - eDropped - eBroken,
```
Comment (tomaz-s1) [inline pkg/client/client.go:464]: Do we have some test for various edge cases which verifies that this value is always >= 0?

File: pkg/client/client.go [hunk 3]
```diff
@@ -406,7 +407,7 @@ func (client *DataSetClient) sendBufferWithRetryPolicy(buf *buffer.Buffer) {
 		} else {
 			err = fmt.Errorf("non recoverable error %w", err)
 			client.onBufferDrop(buf, lastHttpStatus, err)
-			break // exit loop (failed to send buffer)
+			return false // exit loop (failed to send buffer)
```
Comment (tomaz-s1) [inline pkg/client/client.go:410]: Do we ever return `true` on success? I only see us returning false, but I don't see `true` anywhere, but I could have missed it.

File: pkg/client/client.go [hunk 4]
```diff
@@ -431,54 +432,102 @@ func (client *DataSetClient) logStatistics() {
...
+		bAPISent / mb,
+		bAPIAccepted / mb,
+		throughput,
+		perBuffer / mb,
+		successRate,
+		processingInSec,
+		processingDur,
+	}
+
+	return &Statistics{
```
Comment (tomaz-s1) [inline pkg/client/client.go:484]: Yeah, I think that's a nicer / better API. Thanks.

File: pkg/client/statistics.go (hunk 1 above)
Comment (martin-majlis-s1) [reply pkg/client/statistics.go:7]: Fixed! I should have marked it as draft. I have just pushed it before I went home.

File: pkg/client/client.go (hunk 3 above)
Comment (martin-majlis-s1) [reply pkg/client/client.go:410]: We return true here - https://github.com/scalyr/dataset-go/pull/52/files#diff-2895f30116602d8fe6b0545c186d4f81247fa4eaab049dab174c6c91c0036e08R379

File: pkg/client/client.go (hunk 4 above)
Comment (martin-majlis-s1) [reply pkg/client/client.go:484]: My plan was to fix it long time ago but I didn't have the opportunity. :)

File: pkg/client/client.go (hunk 2 above)
Comment (martin-majlis-s1) [reply pkg/client/client.go]: We have test where there is dropped buffer. Broken buffer should never happen.

I have added into few tests some validation that those numbers are correct.

**Comment by martin-majlis-s1** [comment]
And that flaky test still fails. It's interesting that on MacOs it succeeds.

File: pkg/client/statistics.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	// enqueued is number of items that has been accepted for processing
+	enqueued uint64 `mapstructure:"enqueued"`
+	// Processed is number of items that has been successfully processed
+	processed uint64 `mapstructure:"processed"`
+	// dropped is number of items that has been dropped since they couldn't be processed
+	dropped uint64 `mapstructure:"dropped"`
+	// broken is number of items that has been damaged by queue, should be zero
+	broken uint64 `mapstructure:"broken"`
+	// processingTime is duration of the processing
```
Comment (tomaz-s1) [inline pkg/client/statistics.go:17]: I assume that when this value is logged to stdout/stderr, the unit we use is seconds, right? 

It would also help if the comment clarified if all those values are absolute (counters) vs relative (gauge). 

With absolute, I mean counters which are storing the values since the collector start / reload / similar and the value will always be increasing or if it's relative value compared to some previous point in time (e.g. since the last time statistics were logged).

If so, it may make sense to clarify that here.

Comment (martin-majlis-s1) [reply pkg/client/statistics.go:17]: Those numbers are absolute. I can update the comment in 2hrs.

Comment (tomaz-s1) [reply pkg/client/statistics.go:17]: Thanks.

Yeah that would be great. In the mean time, I will go ahead and approve the PR.

Thanks.

**Review by tomaz-s1 — APPROVED** [review]
LGTM
